### PR TITLE
BUG: rolling with datetime ArrowDtype

### LINKED ISSUE #55849: BUG: rolling does not work with timestamp[ns][pyarrow] index type
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
ts = ['2019-01-03 05:11', '2019-01-03 05:23',
      '2019-01-03 05:28', '2019-01-03 05:32',
      '2019-01-03 05:36']
arr = [True, False, False, True, True]
df = pd.DataFrame({'arr': arr, 'ts': ts})
df["ts"] = pd.to_datetime(df["ts"])
df = df.convert_dtypes(dtype_backend="pyarrow").set_index("ts")
print(df.rolling('5min').sum())
```

### Issue Description

Error message: `ValueError: window must be an integer 0 or greater`

### Expected Behavior

No error message

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : a60ad39b4a9febdea9a59d602dad44b1538b0ea5
python              : 3.11.6.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.22621
machine             : AMD64
processor           : Intel64 Family 6 Model 186 Stepping 2, GenuineIntel
byteorder           : little
LC_ALL              : None
LANG                : en
LOCALE              : English_United Kingdom.1252

pandas              : 2.1.2
numpy               : 1.26.0
pytz                : 2023.3.post1
dateutil            : 2.8.2
setuptools          : 68.2.2
pip                 : 23.3.1
Cython              : None
pytest              : None
hypothesis          : None
sphinx              : 7.2.6
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : 2.9.7
jinja2              : 3.1.2
IPython             : 8.17.2
pandas_datareader   : None
bs4                 : 4.12.2
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : None
gcsfs               : None
matplotlib          : 3.8.1
numba               : None
numexpr             : None
odfpy               : None
openpyxl            : 3.1.2
pandas_gbq          : None
pyarrow             : 13.0.0
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.11.3
sqlalchemy          : 2.0.23
tables              : None
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : 2.4.1
pyqt5               : None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #55849 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Additionally renamed _apply_blockwise to _apply_columnwise to fit the actual methodology

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -2514,4 +2515,8 @@ def dtype_to_unit(dtype: DatetimeTZDtype | np.dtype) -> str:
     """
     if isinstance(dtype, DatetimeTZDtype):
         return dtype.unit
+    elif isinstance(dtype, ArrowDtype):
+        if dtype.kind not in "mM":
+            raise ValueError(f"{dtype=} does not have a resolution.")
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimelike.py:2520]: typo here?

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -404,11 +405,12 @@ def _insert_on_column(self, result: DataFrame, obj: DataFrame) -> None:
...
-    def _index_array(self):
+    def _index_array(self) -> npt.NDArray[np.int64] | None:
         # TODO: why do we get here with e.g. MultiIndex?
-        if needs_i8_conversion(self._on.dtype):
-            idx = cast("PeriodIndex | DatetimeIndex | TimedeltaIndex", self._on)
-            return idx.asi8
+        if isinstance(self._on, (PeriodIndex, DatetimeIndex, TimedeltaIndex)):
+            return self._on.asi8
+        elif isinstance(self._on.dtype, ArrowDtype):
+            return self._on.to_numpy(dtype=np.int64)
```
Comment (jbrockmendel) [inline pandas/core/window/rolling.py:413]: comment/assertion about self._on.dtype.kind?

File: pandas/core/window/rolling.py [hunk 3]
```diff
@@ -1871,7 +1875,7 @@ def _validate(self):
         # we allow rolling on a datetimelike index
         if (
             self.obj.empty
-            or isinstance(self._on, (DatetimeIndex, TimedeltaIndex, PeriodIndex))
+            or (isinstance(self._on, PeriodIndex) or self._on.dtype.kind in "Mm")
```
Comment (jbrockmendel) [inline pandas/core/window/rolling.py:1878]: id err on the side of being more explicit about the supported dtypes being ArrowDtype

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/arrays/datetimelike.py:2520]: I don't think so? (If you're referring to `dtype=`, it was intentional to render e.g. `"dtype=int64[pyarrow] does not have a resolution"`)

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:2543]: maybe im misunderstanding how f-strings work.  id expect this to render "int64[pyarrow]= does not have a resolution"

Comment (mroeschke) [reply pandas/core/arrays/datetimelike.py:2543]:
```python
In [2]: import pandas as pd

In [3]: import pyarrow as pa

In [4]: dtype = pd.ArrowDtype(pa.int64())

In [5]: f"{dtype=} does not have a resolution"
Out[5]: 'dtype=int64[pyarrow] does not have a resolution'
```

Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:2543]: neat!

File: pandas/core/window/rolling.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/window/rolling.py:413]: Yup good point. Added a kind check

**Comment by mroeschke** [comment]
Since there's no more comments here going to merge. Can follow up if needed

**Pull request merged**